app/src/main/java/org/wikipedia/analytics/metricsplatform/ArticleEvent.kt

```diff
@@ -0,0 +1,238 @@
+        }
+
+        private fun submitEvent(action: String) {
+            val customData = mapOf(
+                "action" to action,
+                "time_spent_ms" to duration
+            )
+            submitEvent(
+                "android.metrics_platform.article_toolbar_interaction",
+                fragment,
+                customData
+            )
+        }
+    }
+
+    inner class ArticleTocInteraction(
+        private val fragment: PageFragment,
+        private val numSections: Int
+    ) {
+        private var numOpens = 0
+        private var numSectionClicks = 0
+        private var lastScrollStartMillis = 0L
+        private var totalOpenedSec = 0
+
+        fun scrollStart() {
+            numOpens++
+            lastScrollStartMillis = System.currentTimeMillis()
+        }
+
+        fun scrollStop() {
+            if (lastScrollStartMillis == 0L) {
+                return
+            }
+            totalOpenedSec += ((System.currentTimeMillis() - lastScrollStartMillis) / TimeUnit.SECONDS.toMillis(1)).toInt()
+            lastScrollStartMillis = 0
+        }
+
+        fun logClick() {
+            numSectionClicks++
+        }
+
+        fun logEvent() {
+            scrollStop()
+            if (numSections == 0 || numOpens == 0) {
+                return
+            }
+
+            val customData = mapOf(
+                "num_opens" to numOpens,
+                "num_section_clicks" to numSectionClicks,
+                "total_open_sec" to totalOpenedSec,
+                "num_sections" to numSections
+            )
+            submitEvent(
+                "android.metrics_platform.article_toc_interaction",
+                fragment,
+                customData
+            )
+        }
+    }
+
+    inner class ArticleLinkPreviewInteraction(
+        private val fragment: PageFragment,
+        private val source: Int
+    ) {
+        val PROD_LINK_PREVIEW_VERSION = 3
+        fun logLinkClick() {
+            submitEvent("linkclick")
+        }
+
+        fun logNavigate() {
+            submitEvent(if (Prefs.isLinkPreviewEnabled) "navigate" else "disabled")
+        }
+
+        fun logCancel() {
+            submitEvent("cancel")
+        }
+
+        private fun submitEvent(action: String) {
+            val customData = mapOf(
+                "action" to action,
+                "source" to source,
+                "time_spent_ms" to duration,
+                "version" to PROD_LINK_PREVIEW_VERSION
+            )
+            submitEvent(
+                "android.metrics_platform.article_link_preview_interaction",
+                fragment,
+                customData
+            )
+        }
+    }
+
+    private fun submitEvent(
+        eventName: String,
+        fragment: PageFragment,
+        customData: Map<String, Any>
+    ) {
+        val clientMetadata = AndroidPageClientMetadata(fragment)
+        val metricsPlatformClient = MetricsPlatformClientPage(clientMetadata)
```

Comment:
It seems like a `MetricsClient` object is something that should persist for the lifetime of the app process, instead of being recreated upon sending each event. Should this object live as a field in the `WikipediaApp` class?

Reply:
The issue is that page-specific metadata needs to be passed into the `MetricsClient` constructor which is why you'll notice that there are 2 ClientMetadata classes (the page one inherits from the base one). Given that page metadata is dynamic/variable, this was one approach @sharvaniharan and I talked through -- one of the topics i'm hoping to broach in our discussion

Reply:
one thing i'm considering is making a provision in the MP library to update stuff like page metadata -- that might be the better solution

Reply:
the more i think about it, the more it makes sense to provide a mechanism in the library to update metadata -- and keep a single `MetricsClient` object for the lifetime of an app session -- that will simplify things greatly on this end

Reply:
new release published for addressing dynamic data